### PR TITLE
Improve some action for install snapshot and add peer

### PR DESCRIPTION
**Pull request by Linary**
1. Fix bug that relative path need normalize when load snapshot
2. Use zip decompresser since decompress tar file may lead dead loop
3. Move raft.endpoint and raft.group_peers into rest-server.properties
4. Pass raft.endpoint and raft.group_peers from server to core
5. Let RestServer start before GremlinServer
6. Use read-index to ensure the raft log caught up for new node
7. Let add_peer and remove_peer API work in async job
8. Let truncateBackend don't truncate system store

Change-Id: I7de3d98cb51c3b7c32a4f19a0d99ed622f78d331

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java [hunk 1]
```diff
@@ -79,7 +82,16 @@ public void save(SnapshotWriter writer, Closure done,
             // Write snapshot to real directory
             Map<String, String> snapshotDirMaps = this.doSnapshotSave();
             executor.execute(() -> {
+                if (this.compressing.get()) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java:85]: compareAndSet

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java [hunk 2]
```diff
@@ -102,15 +116,23 @@ public boolean load(SnapshotReader reader) {
         Set<String> snapshotDirTars = reader.listFiles();
         LOG.info("The snapshot tar files to be loaded are {}", snapshotDirTars);
         Set<String> snapshotDirs = new HashSet<>();
-        for (String snapshotDirTar : snapshotDirTars) {
-            try {
+        if (this.compressing.get()) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java:119]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java [hunk 3]
```diff
@@ -151,7 +173,13 @@ private void compressSnapshotDir(SnapshotWriter writer,
                                      .toString();
             Checksum checksum = new CRC64();
             try {
-                CompressUtil.compressTar(snapshotDir, outputFile, checksum);
+                LOG.info("Prepare to compress dir {} to {}",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java:176]: '{}'

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java [hunk 4]
```diff
@@ -151,7 +173,13 @@ private void compressSnapshotDir(SnapshotWriter writer,
...
+                         snapshotDir, outputFile);
+                long begin = System.currentTimeMillis();
+                CompressUtil.compressZip(snapshotDir, outputFile, checksum);
+                long end = System.currentTimeMillis();
+                LOG.info("Compressed dir {} to {}, took {} seconds",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java:181]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java [hunk 5]
```diff
@@ -195,7 +223,13 @@ private String decompressSnapshot(SnapshotReader reader,
         Checksum checksum = new CRC64();
         String archiveFile = Paths.get(reader.getPath(), snapshotDirTar)
                                   .toString();
-        CompressUtil.decompressTar(archiveFile, parentPath, checksum);
+        LOG.info("Prepare to decompress snapshot zip {} to {}",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java:226]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java [hunk 6]
```diff
@@ -195,7 +223,13 @@ private String decompressSnapshot(SnapshotReader reader,
...
+                 archiveFile, parentPath);
+        long begin = System.currentTimeMillis();
+        CompressUtil.decompressZip(archiveFile, parentPath, checksum);
+        long end = System.currentTimeMillis();
+        LOG.info("Decompress snapshot zip {} to {}, took {} seconds",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSnapshotFile.java:231]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 7]
```diff
@@ -244,10 +220,19 @@ public static synchronized CoreOptions instance() {
...
+                    "The general rpc timeout for jraft rpc.",
                     positiveInt(),
                     // jraft default value is 5000(ms)
-                    60000
+                    60 * 1000
+            );
+
+    public static final ConfigOption<Integer> RAFT_INSTALL_SNAPSHOT_TIMEOUT =
+            new ConfigOption<>(
+                    "raft.rpc_timeout",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:231]: correct the option name

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java [hunk 8]
```diff
@@ -131,10 +133,15 @@ public void clear() throws BackendException {
     }
 
     @Override
-    public void truncate() {
+    public void truncate(HugeGraph graph) {
         this.checkOpened();
+        HugeConfig config = (HugeConfig) graph.configuration();
+        String systemStoreName = config.get(CoreOptions.STORE_SYSTEM);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java:139]: can split `stores` into multi vars, like schemaStore + systemStore + graphStore

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java [hunk 9]
```diff
@@ -131,10 +133,15 @@ public void clear() throws BackendException {
...
         for (BackendStore store : this.stores.values()) {
-            store.truncate();
+            // Don't truncate system store
+            if (store.store().equals(systemStoreName)) {
+                store.truncate();
+            }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java:144]: it's more reasonable to truncate systemStore + graphStore, can we move meta table from systemStore to schemaStore?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java (hunk 8 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java:139]: tiny comment, address optionally

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 10]
```diff
@@ -106,18 +107,28 @@ public RaftSharedContext(HugeGraphParams params) {
         this.graphStoreName = config.get(CoreOptions.STORE_GRAPH);
         this.systemStoreName = config.get(CoreOptions.STORE_SYSTEM);
         this.stores = new RaftBackendStore[StoreType.ALL.getNumber()];
+
+        // TODO: 依赖了ServerOptions的配置项名，需要打通ServerConfig和CoreConfig
```
[the remark this answers is not in the document]
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:111]: It looks a little complicated.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java (hunk 8 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/AbstractBackendStoreProvider.java:139]: I won't deal with it this time.

**Comment by Linary** [comment]
recheck

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java [hunk 11]
```diff
@@ -219,8 +260,16 @@ public void createSnapshot() {
         StoreCommand command = new StoreCommand(StoreType.GRAPH,
                                                 StoreAction.SNAPSHOT, null);
         RaftStoreClosure closure = new RaftStoreClosure(command);
-        this.context.node().submitAndWait(command, closure);
-        LOG.debug("Graph '{}' has writed snapshot", this.graph());
+        RaftClosure<?> future = this.context.node().submitAndWait(command,
+                                                                  closure);
+        if (future != null) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java:265]: when the future may be null?

Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java:265]: Maybe not. Maybe it's defensive programming

**Review by zyxxoo — DISMISSED** [review]

**Comment by javeme** [comment]
ci error:
```java
2022-05-15 19:04:18 [main] [INFO] c.b.h.b.s.BackendProviderFactory - Opening backend store 'rocksdb' in raft mode for graph 'hugegraph'
2022-05-15 19:04:18 [main] [INFO] c.b.h.b.s.r.RaftBackendStoreProvider - Init raft backend system store
Error: -15 19:04:18 [main] [ERROR] c.b.h.d.HugeGraphServer - HugeRestServer start error: 
java.lang.RuntimeException: GraphFactory could not instantiate this Graph implementation [class com.baidu.hugegraph.auth.HugeFactoryAuthProxy]
	at org.apache.tinkerpop.gremlin.structure.util.GraphFactory.open(GraphFactory.java:84) ~[gremlin-core-3.5.1.jar:3.5.1]
	at org.apache.tinkerpop.gremlin.structure.util.GraphFactory.open(GraphFactory.java:72) ~[gremlin-core-3.5.1.jar:3.5.1]
	at com.baidu.hugegraph.auth.StandardAuthenticator.setup(StandardAuthenticator.java:121) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.auth.HugeAuthenticator.loadAuthenticator(HugeAuthenticator.java:148) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.core.GraphManager.<init>(GraphManager.java:96) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.server.ApplicationConfig$GraphManagerFactory$1.onEvent(ApplicationConfig.java:112) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at org.glassfish.jersey.server.internal.monitoring.CompositeApplicationEventListener.onEvent(CompositeApplicationEventListener.java:49) ~[jersey-server-3.0.3.jar:?]
	at org.glassfish.jersey.server.internal.monitoring.MonitoringContainerListener.onStartup(MonitoringContainerListener.java:56) ~[jersey-server-3.0.3.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.onStartup(ApplicationHandler.java:711) ~[jersey-server-3.0.3.jar:?]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.start(GrizzlyHttpContainer.java:330) ~[jersey-container-grizzly2-http-3.0.3.jar:?]
	at org.glassfish.grizzly.http.server.HttpHandlerChain.start(HttpHandlerChain.java:376) ~[grizzly-http-server-3.0.1.jar:3.0.1]
	at org.glassfish.grizzly.http.server.HttpServer.setupHttpHandler(HttpServer.java:268) ~[grizzly-http-server-3.0.1.jar:3.0.1]
	at org.glassfish.grizzly.http.server.HttpServer.start(HttpServer.java:245) ~[grizzly-http-server-3.0.1.jar:3.0.1]
	at com.baidu.hugegraph.server.RestServer.start(RestServer.java:68) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.server.RestServer.start(RestServer.java:175) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.dist.HugeRestServer.start(HugeRestServer.java:34) ~[hugegraph-dist-0.13.0.jar:?]
	at com.baidu.hugegraph.dist.HugeGraphServer.<init>(HugeGraphServer.java:62) ~[hugegraph-dist-0.13.0.jar:?]
	at com.baidu.hugegraph.dist.HugeGraphServer.main(HugeGraphServer.java:121) ~[hugegraph-dist-0.13.0.jar:?]
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_332]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_332]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_332]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_332]
	at org.apache.tinkerpop.gremlin.structure.util.GraphFactory.open(GraphFactory.java:80) ~[gremlin-core-3.5.1.jar:3.5.1]
	... 17 more
Caused by: java.lang.IllegalStateException: Please ensure init raft context
	at com.google.common.base.Preconditions.checkState(Preconditions.java:531) ~[guava-25.1-jre.jar:?]
	at com.baidu.hugegraph.util.E.checkState(E.java:68) ~[hugegraph-common-2.1.2.jar:2.1.2.0]
	at com.baidu.hugegraph.backend.store.raft.RaftBackendStoreProvider.context(RaftBackendStoreProvider.java:74) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.backend.store.raft.RaftBackendStoreProvider.loadSystemStore(RaftBackendStoreProvider.java:143) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.StandardHugeGraph.loadSystemStore(StandardHugeGraph.java:466) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.StandardHugeGraph.access$1300(StandardHugeGraph.java:113) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.StandardHugeGraph$StandardHugeGraphParams.loadSystemStore(StandardHugeGraph.java:1185) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.task.ServerInfoManager.listenChanges(ServerInfoManager.java:109) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.task.ServerInfoManager.<init>(ServerInfoManager.java:85) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.task.StandardTaskScheduler.<init>(StandardTaskScheduler.java:106) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.task.TaskManager.addScheduler(TaskManager.java:92) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.StandardHugeGraph.<init>(StandardHugeGraph.java:218) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.HugeFactory.open(HugeFactory.java:91) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.HugeFactory.open(HugeFactory.java:69) ~[hugegraph-core-0.13.0.jar:0.13.0.0]
	at com.baidu.hugegraph.auth.HugeFactoryAuthProxy.open(HugeFactoryAuthProxy.java:83) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_332]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_332]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_332]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_332]
	at org.apache.tinkerpop.gremlin.structure.util.GraphFactory.open(GraphFactory.java:80) ~[gremlin-core-3.5.1.jar:3.5.1]
	... 17 more
The operation timed out(60s) when attempting to connect to http://127.0.0.1:8082/graphs
```

**Comment by Linary** [comment]
+2, Thanks!

**Review by zyxxoo — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Comment by wangyao2016** [comment]
With this PR, I cannot  start hugegraph successful. Err:

`2022-05-17 18:37:28 [main] [INFO] c.b.h.s.RestServer - RestServer starting...
2022-05-17 18:37:30 [main] [INFO] c.b.h.u.ConfigUtil - Scanning option 'graphs' directory './conf/graphs'
2022-05-17 18:37:30 [main] [INFO] c.b.h.b.c.Cache - Init RamCache for 'users-hugegraph' with capacity 10240
2022-05-17 18:37:30 [main] [INFO] c.b.h.b.c.Cache - Init RamCache for 'users_pwd-hugegraph' with capacity 10240
2022-05-17 18:37:30 [main] [INFO] c.b.h.b.c.Cache - Init RamCache for 'token-hugegraph' with capacity 10240
2022-05-17 18:37:30 [main] [INFO] c.b.h.s.RestServer - Graph 'hugegraph' was successfully configured via './conf/graphs/hugegraph.properties'
2022-05-17 18:37:30 [main] [INFO] c.b.h.r.RpcServer - RpcServer config is empty, skip starting RpcServer
2022-05-17 18:37:30 [main] [ERROR] c.b.h.d.HugeGraphServer - HugeRestServer start error: 
java.lang.NullPointerException: null
	at com.baidu.hugegraph.testutil.Whitebox.getInternalState(Whitebox.java:69) ~[hugegraph-common-2.1.2.jar:2.1.2.0]
	at com.baidu.hugegraph.core.GraphManager.remotingRpcServer(GraphManager.java:301) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.core.GraphManager.waitGraphsReady(GraphManager.java:373) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.core.GraphManager.<init>(GraphManager.java:112) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.server.ApplicationConfig$GraphManagerFactory$1.onEvent(ApplicationConfig.java:130) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at org.glassfish.jersey.server.internal.monitoring.CompositeApplicationEventListener.onEvent(CompositeApplicationEventListener.java:49) ~[jersey-server-3.0.3.jar:?]
	at org.glassfish.jersey.server.internal.monitoring.MonitoringContainerListener.onStartup(MonitoringContainerListener.java:56) ~[jersey-server-3.0.3.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.onStartup(ApplicationHandler.java:711) ~[jersey-server-3.0.3.jar:?]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.start(GrizzlyHttpContainer.java:330) ~[jersey-container-grizzly2-http-3.0.3.jar:?]
	at org.glassfish.grizzly.http.server.HttpHandlerChain.start(HttpHandlerChain.java:376) ~[grizzly-http-server-3.0.1.jar:3.0.1]
	at org.glassfish.grizzly.http.server.HttpServer.setupHttpHandler(HttpServer.java:268) ~[grizzly-http-server-3.0.1.jar:3.0.1]
	at org.glassfish.grizzly.http.server.HttpServer.start(HttpServer.java:245) ~[grizzly-http-server-3.0.1.jar:3.0.1]
	at com.baidu.hugegraph.server.RestServer.start(RestServer.java:73) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.server.RestServer.start(RestServer.java:180) ~[hugegraph-api-0.13.0.jar:0.69.0.0]
	at com.baidu.hugegraph.dist.HugeRestServer.start(HugeRestServer.java:34) ~[hugegraph-dist-0.13.0.jar:?]
	at com.baidu.hugegraph.dist.HugeGraphServer.<init>(HugeGraphServer.java:62) ~[hugegraph-dist-0.13.0.jar:?]
	at com.baidu.hugegraph.dist.HugeGraphServer.main(HugeGraphServer.java:121) ~[hugegraph-dist-0.13.0.jar:?]`

I found [serverConfig.getServer()](https://github.com/apache/incubator-hugegraph/commit/8609968e6648d69189b6ea6d37b68b81e245e4dc#diff-7a0b04965824fe2f392d66bc9cb0addb6b059d81142c3117f8ff6c390b5e4e03R299)  was null. Actually ServerConfig is `[protocol=bolt, port=8091, host=127.0.0.1]`


Can you take a look？Thanks. @Linary

**Pull request merged**